objectserver/auditor.go

```diff
@@ -143,6 +217,109 @@ func auditHash(hashPath string, skipMd5 bool) (bytesProcessed int64, err error)
 	return bytesProcessed, nil
 }
 
+func quarantineShard(db *IndexDB, hash string, shard int, timestamp int64, metabytes []byte, nursery bool) error {
+	shardPath, err := db.WholeObjectPath(hash, shard, timestamp, nursery)
+	if err != nil {
+		return err
+	}
+	shardName := filepath.Base(shardPath)
+	objsDir := filepath.Dir(filepath.Dir(filepath.Dir(shardPath)))
+	driveDir := filepath.Dir(objsDir)
+	quarantineDir := filepath.Join(driveDir, "quarantined", filepath.Base(objsDir))
+	if err := os.MkdirAll(quarantineDir, 0755); err != nil {
+		return err
+	}
+	metadata := map[string]string{}
+	if err = json.Unmarshal(metabytes, &metadata); err != nil {
+		return fmt.Errorf("Error decoding metadata: %s", err)
+	}
+	name, ok := metadata["name"]
+	if !ok {
+		return fmt.Errorf("Metadata missing name: %s", string(metabytes))
+	}
+	fName := filepath.Join(quarantineDir, shardName)
+	f, err := os.OpenFile(fName, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
+	}
+	_, err = f.WriteString(name)
+	f.Close()
+	if err != nil {
+		if rmErr := os.Remove(fName); rmErr != nil {
+			return rmErr
+		}
+		return err
+	}
+	return db.Remove(hash, shard, timestamp, nursery)
+}
+
+// auditDB.  Runs auditFunc on all objects in the given DB.
+func (a *Auditor) auditDB(dbpath string, objRing ring.Ring) {
+	policyDir := filepath.Dir(dbpath)
+	path := filepath.Join(policyDir, "hec")
+	temppath := filepath.Join(filepath.Dir(policyDir), "tmp")
+	ringPartPower := bits.Len64(objRing.PartitionCount() - 1)
+	zapLogger, ok := a.logger.(*zap.Logger)
+	if !ok {
+		a.logger.Error("Logger type assertion failed")
+		zapLogger = zap.L()
+	}
+	db, err := NewIndexDB(dbpath, path, temppath, ringPartPower, 1, 32, zapLogger)
```

Comment:
This isn't going to work with the new changes where db_part_power and subdirs are policy config options.